양문규/Week25/도서관.py

```diff
@@ -0,0 +1,79 @@
+import sys
+from enum import Enum
+from collections import deque
+from heapq import heappush, heappop, heapify
+import math
+
+
+# 가장 먼 거리
+max_dist = 0
+
+# 음수 저장
+list_left = []
+# 양수 저장
+list_right = []
+
+# 입력 받기
+N, M = map(int, input().split())
+
+line = list(map(int, input().split()))
+for n in line:
+    if n < 0:
+        # 절대값을 적용한 값을 넣음)
+        list_left.append(abs(n))
+        max_dist = max(max_dist, abs(n))
+    else:
+        list_right.append(n)
+        max_dist = max(max_dist, n)
+
+# 내림차순으로 정렬
+list_left.sort()
+list_right.sort()
+
+
+def return_books_left():
+    answer = list_left[-1] * 2
+
+    i = M
+    while i > 0 and len(list_left):
+        list_left.pop()
+        i -= 1
+
+    return answer
+
+
+def return_books_right():
+    answer = list_right[-1] * 2
+
+    i = M
+    while i > 0 and len(list_right):
+        list_right.pop()
+        i -= 1
+
+    return answer
+
+
+def solution():
+    answer = 0
+
+    while len(list_left) and len(list_right):
```

Comment:
```suggestion
    while list_left and list_right:
```

파이썬 empty 상태 확인 할 때 len() 호출 없이 가능하지 않나요?

Reply:
저도 질문을 보고 AI에게 물어보니 둘 다 모두 가능하다네요.
파이썬은 빈 배열도 Falsy로 취급한다는군요! 하나배웠습니다. 👍

Reply:
<img width="761" alt="image" src="https://github.com/user-attachments/assets/928664cb-9645-49a2-bd05-da8aff789bda">
궁금해서 찾아보니 len으로 접근해도 O(1)만큼 걸린다하니, len을 쓴다고 성능상 불이익을 보진 않을 것 같네요. 취향을 타는 부분일 수 있겠군요!

Reply:
파이썬의 경우는 잘 모르지만 Java 를 예시로 보면

```java
char[] chars = new char[] {'H', 'e', 'l', 'l', 'o', ' ', 'W', 'o', 'r', 'l', 'd'};

chars.length;  //  배열이 갖는 길이 필드 값
```

배열의 경우 선언과 동시에 필드 값으로 capacity 값이 상수로 초기화 되어서 배열의 길이가 정해져서 길이를 다시 측정하지 않아도 됩니다. (그래서 배열은 선언과 동시에 초기화 혹은 길이가 정해져야 함)

```java
char[] chars = new char[10];  // example) 길이와 같이 초기화
```

제 생각에는 파이썬의 경우도 비효율적으로 매번 호출할 때마다 측정하지 않고, 내부에 갖고 있는 길이 정보를 리턴 해줄 것 같네요.
말 그대로 취향 차이인 것 같습니다. 👍

Reply:
궁금해서 조금 더 살펴 봤는데

```c
typedef struct {
    Py_ssize_t allocated;
    PyObject *ob_item[];
} _PyListArray;

static Py_ssize_t
list_length(PyObject *a)
{
    return PyList_GET_SIZE(a);
}
```

PyList_GET_SIZE 가 사이즈를 반환 해주는 코드인가 봅니다 ㅎ...
#define PyList_GET_SIZE 되어있는 헤더 파일은 어디 있는지 모르겠네요.

단순히 길이만 반환 해주는거 같습니다 ~

Reply:
오.. 여기까지 살펴보시다니..!
`ssize`는 뭘까요 ㅋ
`ㅅ..사이즈`?

Reply:
GPT 가 위치 다 알고 있네요 ㄷㄷ.. [[헤더 위치]](https://github.com/python/cpython/blob/main/Include/cpython/listobject.h)

앞에 s는 signed / sign / short 중 하나가 아닐까 싶은데 c 언어 오래만에 보니까 쉽지 않네요 ㅋㅋ...